### PR TITLE
Add support for rolling snapshot zone 

### PR DESCRIPTION
**Pull request by weilewei**
This PR adds support for rolling snapshot zone asynchronously. **This PR should be reviewed after #48 is reviewed or merged.**
- adds `RollSnapshotZone` and `WriteSnapshot` functions, and added roll and write logics where needed.
- roll snapshot if needed when rolling log zone. package the snapshot zone rolling process to a background worker.

With this change, 1) the db_bench can run; 2) zenfs_test can run successfully when the number of written files is set to 13 [highest available open zone]; 3) zenfs_metazone_rollover_test.cc can be passed.

TODO:
- since the background worker refactor is WIP by @levisonchen, some backgroundWorker related code still needs refactor later accordingly.

### CODE REVIEW

File: fs/fs_zenfs.cc [hunk 1]
```diff
@@ -302,8 +319,72 @@ IOStatus ZenFS::RollMetaZone() {
...
+  std::string super_string;
+  superblock_->EncodeTo(&super_string);
+  s = meta_log_->AddRecord(super_string);
+  if (!s.ok()) {
+    Error(logger_,
+          "Could not write super block when rolling to a new meta zone");
+    return IOStatus::IOError("Failed writing a new superblock");
+  }
+
+  BackgroundWorker bg_worker;
```
Comment (royguo) [inline fs/fs_zenfs.cc]: I think we should have a global bg_worker, this is just for temporary test, right? @weilewei @levisonchen

File: fs/fs_zenfs.cc [hunk 2]
```diff
@@ -302,8 +319,72 @@ IOStatus ZenFS::RollMetaZone() {
...
+      old_meta_zone->Reset();
+    }
+
+    // TODO: 返回值需要不是int
+    return s.ok();
+  };
+
+  // TODO: arg is a placeholder to satisfy SubmitJob signature
+  int arg = 42;
+  bg_worker.SubmitJob(task, (void*)&arg);
```
Comment (royguo) [inline fs/fs_zenfs.cc]: This is not good-looking... do you think we can remove the args here? I think lambda-based task is enough. @levisonchen

File: fs/fs_zenfs.h [hunk 3]
```diff
@@ -152,6 +152,7 @@ class ZenFS : public FileSystemWrapper {
   void LogFiles();
   void ClearFiles();
   IOStatus WriteSnapshotLocked(ZenMetaLog* meta_log);
+  IOStatus WriteSnapshotInSnapshotZone(ZenMetaLog* meta_snapshot_log);
```
Comment (royguo) [inline fs/fs_zenfs.h]: A minor suggestion, we can just use `WriteSnapshot` here

File: test/backgroundWorker_test.cc [hunk 4]
```diff
@@ -2,8 +2,8 @@
 #include <fcntl.h>
 #include <gflags/gflags.h>
 #include <rocksdb/file_system.h>
-#include <rocksdb/plugin/zenfs/fs/fs_zenfs.h>
-#include <rocksdb/plugin/zenfs/fs/zbd_zenfs.h>
+#include <third-party/zenfs/fs/fs_zenfs.h>
+#include <third-party/zenfs/fs/zbd_zenfs.h>
```
Comment (royguo) [inline test/backgroundWorker_test.cc]: IIRC, I put the `third-party/zenfs` in CMake's include_directories, maybe we should use `#include "fs/fs_zenfs.h"` here, please have a look. And for the coding style, we only use `<>` for system provided or third-party provided headers, for self-contained headers we use `“ ”`.

**Comment by levichen94** [comment]
> Do not review and merge. PR for discussion only.
> 
> Questions:
> 
> * current BackgroundWorker::SubmitJob only allows job function of return type of int. But we might need to allow IOStatus type. Also input argument type is restricted to void*.
> * any example for handling space identification for a zone?

I'll give some examples in our meeting.

File: fs/fs_zenfs.cc (hunk 2 above)
Comment (levichen94) [reply fs/fs_zenfs.cc]: We could structurize that, but we also needs arguments for jobs, it's just a change of coding style.

File: fs/fs_zenfs.cc (hunk 1 above)
Comment (levichen94) [reply fs/fs_zenfs.cc]: Bg_worker as ZBD's member will be better.

File: fs/io_zenfs.cc [hunk 5]
```diff
@@ -557,7 +557,7 @@ IOStatus ZonedWritableFile::FlushBuffer() {
   if (pad_sz) memset((char*)buffer + buffer_pos, 0x0, pad_sz);
 
   wr_sz = buffer_pos + pad_sz;
-  s = zoneFile_->Append((char*)buffer, wr_sz, buffer_pos, true);
+  s = zoneFile_->Append((char*)buffer, wr_sz, buffer_pos);
```
Comment (royguo) [inline fs/io_zenfs.cc:559]: Add `false` & comments TODB

File: fs/zbd_zenfs.cc [hunk 6]
```diff
@@ -39,16 +39,16 @@
...
+ * to roll the log safely. One extra
  * is allocated to cover for one zone going offline.
  */
-#define ZENFS_META_ZONES (4)
+#define ZENFS_OP_LOG_ZONES (2)
 
 /* Number of reserved zones for metadata snapshot
  */
-#define ZENFS_SNAPSHOT_ZONES (2)
+#define ZENFS_snapshot_zones_ (2)
```
Comment (royguo) [inline fs/zbd_zenfs.cc]: Naming correcting

**Pull request merged**